File: src/core/stats.ts
/**
 * Reviewer statistics arithmetic (spec §13.1). Descriptive thresholds only —
 * output never feeds back into enforcement.
 */

export interface DecisionPoint {
  reviewer: string;
  received_ms: number;
  seq: number;          // audit sequence of the DECISION_RECORDED event
  approved: boolean;    // approve or release
  latency_ms: number;   // received_ms - first_view_ms
}

export interface ReviewerStats {
  reviewer: string;
  decisions: number;
  approved: number;
  rejected: number;
  approval_bps: number | null;
  median_latency_ms: number | null;
  subsecond_bps: number | null;
  prior_approval_bps: number | null;
  prior_median_latency_ms: number | null;
  flag: "INSUFFICIENT_DATA" | "NONE" | "HIGH_APPROVAL_FAST" | "RISING_APPROVAL_FALLING_LATENCY";
}

export interface StatsInput {
  from_ms: number;
  to_ms: number;
  /** Committed accepted decisions (any seq source), all reviewers. */
  decisions: DecisionPoint[];
  /** Current policy roster (for union with interval-observed reviewers). */
  roster: string[];
}

function lowerMedian(sorted: number[]): number {
  return sorted[Math.floor((sorted.length - 1) / 2)]!;
}

function windowStats(w: DecisionPoint[]): {
  decisions: number; approved: number; rejected: number;
  approval_bps: number | null; median_latency_ms: number | null; subsecond_bps: number | null;
} {
  const n = w.length;
  if (n === 0) {
    return { decisions: 0, approved: 0, rejected: 0, approval_bps: null, median_latency_ms: null, subsecond_bps: null };
  }
  const approved = w.filter((d) => d.approved).length;
  const lat = w.map((d) => d.latency_ms).sort((a, b) => a - b);
  const sub = w.filter((d) => d.latency_ms < 1000).length;
  return {
    decisions: n,
    approved,
    rejected: n - approved,
    approval_bps: Math.floor((10000 * approved) / n),
    median_latency_ms: lowerMedian(lat),
    subsecond_bps: Math.floor((10000 * sub) / n),
  };
}

export function computeReviewerStats(input: StatsInput): ReviewerStats[] {
  const inInterval = input.decisions.filter(
    (d) => d.received_ms >= input.from_ms && d.received_ms < input.to_ms,
  );
  const observed = new Set(inInterval.map((d) => d.reviewer));
  const union = new Set([...input.roster, ...observed]);
  const reviewers = [...union].sort();

  const out: ReviewerStats[] = [];
  for (const reviewer of reviewers) {
    const mine = inInterval
      .filter((d) => d.reviewer === reviewer)
      .sort((a, b) => a.received_ms - b.received_ms || a.seq - b.seq);
    const newest40 = mine.slice(-40);
    const current = newest40.slice(-20);
    const prior = newest40.slice(0, Math.max(0, newest40.length - 20)).slice(-20);

    const cur = windowStats(current);
    const hasPrior = prior.length === 20;
    const pri = hasPrior ? windowStats(prior) : null;

    let flag: ReviewerStats["flag"];
    if (cur.decisions < 20) {
      flag = "INSUFFICIENT_DATA";
    } else if (cur.approval_bps! >= 9500 && cur.median_latency_ms! < 1000) {
      flag = "HIGH_APPROVAL_FAST";
    } else if (
      pri !== null &&
      cur.approval_bps! - pri.approval_bps! >= 1000 &&
      2 * cur.median_latency_ms! <= pri.median_latency_ms!
    ) {
      flag = "RISING_APPROVAL_FALLING_LATENCY";
    } else {
      flag = "NONE";
    }

    out.push({
      reviewer,
      decisions: cur.decisions,
      approved: cur.approved,
      rejected: cur.rejected,
      approval_bps: cur.approval_bps,
      median_latency_ms: cur.median_latency_ms,
      subsecond_bps: cur.subsecond_bps,
      prior_approval_bps: pri ? pri.approval_bps : null,
      prior_median_latency_ms: pri ? pri.median_latency_ms : null,
      flag,
    });
  }
  return out;
}

File: python/lextier/stats.py
"""Reviewer statistics arithmetic (spec §13.1) — parity with
src/core/stats.ts. Descriptive thresholds only; output never feeds back into
enforcement."""


def _lower_median(srt):
    return srt[(len(srt) - 1) // 2]


def _window_stats(w):
    n = len(w)
    if n == 0:
        return {"decisions": 0, "approved": 0, "rejected": 0,
                "approval_bps": None, "median_latency_ms": None, "subsecond_bps": None}
    approved = sum(1 for d in w if d["approved"])
    lat = sorted(d["latency_ms"] for d in w)
    sub = sum(1 for d in w if d["latency_ms"] < 1000)
    return {
        "decisions": n,
        "approved": approved,
        "rejected": n - approved,
        "approval_bps": (10000 * approved) // n,
        "median_latency_ms": _lower_median(lat),
        "subsecond_bps": (10000 * sub) // n,
    }


def compute_reviewer_stats(from_ms: int, to_ms: int, decisions, roster):
    """decisions: [{reviewer, received_ms, seq, approved, latency_ms}];
    roster: current policy reviewers. Returns one record per reviewer in the
    roster ∪ interval-observed union, sorted ascending by PrincipalId."""
    in_interval = [d for d in decisions if from_ms <= d["received_ms"] < to_ms]
    observed = {d["reviewer"] for d in in_interval}
    reviewers = sorted(set(roster) | observed)

    out = []
    for reviewer in reviewers:
        mine = sorted((d for d in in_interval if d["reviewer"] == reviewer),
                      key=lambda d: (d["received_ms"], d["seq"]))
        newest40 = mine[-40:]
        current = newest40[-20:]
        prior = newest40[:max(0, len(newest40) - 20)][-20:]

        cur = _window_stats(current)
        pri = _window_stats(prior) if len(prior) == 20 else None

        if cur["decisions"] < 20:
            flag = "INSUFFICIENT_DATA"
        elif cur["approval_bps"] >= 9500 and cur["median_latency_ms"] < 1000:
            flag = "HIGH_APPROVAL_FAST"
        elif (pri is not None
              and cur["approval_bps"] - pri["approval_bps"] >= 1000
              and 2 * cur["median_latency_ms"] <= pri["median_latency_ms"]):
            flag = "RISING_APPROVAL_FALLING_LATENCY"
        else:
            flag = "NONE"

        out.append({
            "reviewer": reviewer,
            "decisions": cur["decisions"],
            "approved": cur["approved"],
            "rejected": cur["rejected"],
            "approval_bps": cur["approval_bps"],
            "median_latency_ms": cur["median_latency_ms"],
            "subsecond_bps": cur["subsecond_bps"],
            "prior_approval_bps": pri["approval_bps"] if pri else None,
            "prior_median_latency_ms": pri["median_latency_ms"] if pri else None,
            "flag": flag,
        })
    return out
